plugins/http/static/etag.go

```diff
@@ -0,0 +1,71 @@
+package static
+
+import (
+	"hash/crc32"
+	"io"
+	"net/http"
+	"os"
+	"unsafe"
+
+	httpConfig "github.com/spiral/roadrunner/v2/plugins/http/config"
+)
+
+const etag string = "Etag"
+
+// weak Etag prefix
+var weakPrefix = []byte(`W/`)
+
+// CRC32 table
+var crc32q = crc32.MakeTable(0x48D90782)
+
+func SetEtag(cfg *httpConfig.Static, f *os.File, w http.ResponseWriter) {
+	// read the file content
+	body, err := io.ReadAll(f)
+	if err != nil {
+		return
+	}
+
+	// skip for 0 body
+	if len(body) == 0 {
+		return
+	}
+
+	// preallocate
+	calculatedEtag := make([]byte, 0, 64)
+
+	// write weak
+	if cfg.Weak {
```

Comment:
I don't understand the purpose of `Weak` when it only adds the `W/` prefix but does not change the way the ETag is calculated. IMO crc32 is a fairly strong ETag.

My understanding is that `W/` is used when the ETag has more in-depth knowledge of the content of the asset. For example, if the current date appears in the asset, you might wish to keep the ETag even when the date changes.

Reply:
You can read the RFC: https://tools.ietf.org/html/rfc7232#section-2.1 to get understandings. Also, the hash algorithm doesn't matter here. Historically MD5 hash was used for the `Etag`. Also, have a look at the `If-Match` and `If-None-Match` for the `Etag`.
Also, to understand better, take a look at the RFC 7232 section 6.

Reply:
As for CRC32 used instead of MD5 is because on modern (5-10 years I guess) CPUs you may have an advantage in computation for the big HTTP bodies using `PCLMULQDQ` CPU instruction.

Reply:
`W/` after set processed by the `http/fs` which carefully follow `RFC 7232` for the weak and strong `Etag`s.

Reply:
I understand, that it might be too long to read the whole RFC, this is few lines of `http/fs` code and you'll see the difference (not all for sure):
```go
// etagStrongMatch reports whether a and b match using strong ETag comparison.
// Assumes a and b are valid ETags.
func etagStrongMatch(a, b string) bool {
	return a == b && a != "" && a[0] == '"'
}

// etagWeakMatch reports whether a and b match using weak ETag comparison.
// Assumes a and b are valid ETags.
func etagWeakMatch(a, b string) bool {
	return strings.TrimPrefix(a, "W/") == strings.TrimPrefix(b, "W/")
}
```

Reply:
That was a really good question :)
Generally, as for the `Weak` vs `Strong` **calculation**, from the `static` plugin POV they are the same. For example, I can calculate `weak` Etag only based on the size of the `http` Body, and strong Etag value based on the binary content. But, I've tested the CRC32 performance on the big bodies and it was around 300-500ns for the kilobytes. And I decided to uniform the value of the Etag. But, for sure, If I see performance degradation here, I'll use a way faster algorithm for the `Weak` Etag value.